src/engineio/transports.rs

```diff
@@ -0,0 +1,129 @@
+            transport_type: Arc::new(RwLock::new(TransportTypes::Polling)),
+        }
+    }
+
+    pub(super) fn upgrade_websocket(&mut self, address: String) -> Result<()> {
+        if self.websocket_secure.lock()?.is_none() {
+            *self.websocket.lock()? = Some(WebsocketTransport::new(address));
+            self.websocket.lock()?.as_ref().unwrap().probe()?;
+            *self.transport_type.write()? = TransportTypes::Websocket;
+            Ok(())
+        } else {
+            Err(Error::TransportExists())
+        }
+    }
+
+    pub(super) fn upgrade_websocket_secure(&mut self, address: String) -> Result<()> {
+        if self.websocket_secure.lock()?.is_none() {
+            *self.websocket_secure.lock()? = Some(WebsocketSecureTransport::new(
+                address,
+                self.tls_config.lock()?.clone(),
+            ));
+            self.websocket_secure.lock()?.as_ref().unwrap().probe()?;
+            *self.transport_type.write()? = TransportTypes::WebsocketSecure;
+            Ok(())
+        } else {
+            Err(Error::TransportExists())
+        }
+    }
+}
+
+impl Transport for Transports {
+    fn emit(&self, address: String, data: Bytes, is_binary_att: bool) -> Result<()> {
+        match &*self.transport_type.read()? {
+            TransportTypes::Websocket => {
+                self.websocket
+                    .lock()?
+                    .as_ref()
+                    .unwrap()
+                    .emit(address, data, is_binary_att)
+            }
+            TransportTypes::WebsocketSecure => self
+                .websocket_secure
+                .lock()?
+                .as_ref()
+                .unwrap()
+                .emit(address, data, is_binary_att),
+            TransportTypes::Polling => {
+                self.polling
+                    .lock()?
+                    .as_ref()
+                    .unwrap()
+                    .emit(address, data, is_binary_att)
+            }
+        }
+    }
+
+    fn poll(&mut self, address: String) -> Result<Bytes> {
+        match &*self.transport_type.read()? {
+            TransportTypes::Websocket => self
+                .websocket_secure
+                .lock()?
+                .as_mut()
+                .unwrap()
+                .poll(address),
+            TransportTypes::WebsocketSecure => {
+                self.websocket.lock()?.as_mut().unwrap().poll(address)
+            }
+            TransportTypes::Polling => self.polling.lock()?.as_mut().unwrap().poll(address),
+        }
+    }
+
+    fn name(&self) -> Result<String> {
+        match &*self.transport_type.read()? {
+            TransportTypes::Websocket => self.websocket_secure.lock()?.as_ref().unwrap().name(),
+            TransportTypes::WebsocketSecure => self.websocket.lock()?.as_ref().unwrap().name(),
+            TransportTypes::Polling => self.polling.lock()?.as_ref().unwrap().name(),
+        }
+    }
```

Comment:
Do you think that it might be nicer to maybe have a `to_string` method on `TransportType`? As fat as I can see this would also make the `name` method on `Transport` obsolete.

Reply:
Hmmm, isn't to_string more of I want to see this object as a human readable string rather than seeing just it's type? Changed this function to return the string rater than having it in the trait.

Reply:
Yeah that looks good!